### PR TITLE
Add ability to combine two QuantumWorlds

### PR DESCRIPTION
**Pull request by dstrain115**
- Combines the QuantumObjects of both worlds,
in addition to the post-selection criteria.

- This functionality is needed for the Quantum RPG.

### CODE REVIEW

**Comment by losos0** [comment]
Please move whitespace changes to a separate commit. It's hard to find the relevant diffs

**Comment by dstrain115** [comment]
Whitespace changes merged, PTAL

File: unitary/alpha/quantum_world.py [hunk 1]
```diff
@@ -52,6 +52,16 @@ def add_object(self, obj: QuantumObject):
...
         obj.initial_effect()
 
+    def combine_with(self, other_world: "QuantumWorld"):
+        if self.used_object_keys.intersection(other_world.used_object_keys):
+            raise ValueError("Cannot combine two worlds with overlapping object keys")
+        self.objects.extend(other_world.objects)
+        self.used_object_keys.update(other_world.used_object_keys)
+        self.ancilla_names.update(other_world.ancilla_names)
+        self.post_selection.update(other_world.post_selection)
+        for op in other_world.circuit.all_operations():
```
Comment (losos0) [inline unitary/alpha/quantum_world.py:62]: You can write `self.circuit += other.circuit`

File: unitary/alpha/quantum_world.py [hunk 2]
```diff
@@ -52,6 +52,16 @@ def add_object(self, obj: QuantumObject):
...
+            self.circuit.append(op)
+
```
Comment (losos0) [inline unitary/alpha/quantum_world.py:64]: Does `self.effect_history` need to be updated? Cleared maybe.

File: unitary/alpha/quantum_world.py [hunk 3]
```diff
@@ -52,6 +52,16 @@ def add_object(self, obj: QuantumObject):
         obj.world = self
         obj.initial_effect()
 
+    def combine_with(self, other_world: "QuantumWorld"):
```
Comment (losos0) [inline unitary/alpha/quantum_world.py:55]: What's the status of other_world after the operation? Is it still valid to use?

There should be a comment explaining which object contains the combined world after this is called.

File: unitary/alpha/quantum_world.py (hunk 2 above)
Comment (dstrain115) [reply unitary/alpha/quantum_world.py:98]: It's now cleared.

File: unitary/alpha/quantum_world.py (hunk 3 above)
Comment (dstrain115) [reply unitary/alpha/quantum_world.py:78]: Added a docstring to explain this.  I also realized the QuantumWorld class is missing a docstring, so added that as well.

I also now explicitly clear the other_world.

File: unitary/alpha/quantum_world.py [hunk 4]
```diff
@@ -66,7 +110,8 @@ def undo_last_effect(self):
         Note that pop() is considered to be an effect for the purposes
         of this call.
         """
-        self.circuit, self.post_selection = self.effect_history.pop()
+        if self.effect_history:
```
Comment (losos0) [inline unitary/alpha/quantum_world.py:113]: Throwing an exception seems better if there is no last effect.

**Review by losos0 — APPROVED** [review]

File: unitary/alpha/quantum_world.py (hunk 4 above)
Comment (dstrain115) [reply unitary/alpha/quantum_world.py]: Done.

**Pull request merged**